File: internal/bot/rl_encoder.go
package bot

import "github.com/Mgrdich/TermChess/internal/engine"

const (
	pieceChannels        = 12
	currentPosChannels   = 18
	numHistoryPositions  = 4
	numChannels          = currentPosChannels + numHistoryPositions*pieceChannels // 66
	boardSize            = 8
	encodingSize         = numChannels * boardSize * boardSize // 4224
	policySize           = 4096                                // 64 * 64 from-to squares
)

// encodeBoard converts a chess board position and its history into a 66-channel
// float32 tensor matching the Python training encoder exactly.
// Returns a flat []float32 of length 4224 in [channel, rank, file] order.
//
// Channel layout:
//
//	 0-5:  White pieces (Pawn, Knight, Bishop, Rook, Queen, King) — current position
//	 6-11: Black pieces — current position
//	 12:   Side to move (1.0 if White to move)
//	 13:   White kingside castling available
//	 14:   White queenside castling available
//	 15:   Black kingside castling available
//	 16:   Black queenside castling available
//	 17:   En passant file (column filled with 1.0 if en passant available)
//	 18-29: Piece planes for position t-1 (most recent history)
//	 30-41: Piece planes for position t-2
//	 42-53: Piece planes for position t-3
//	 54-65: Piece planes for position t-4
func encodeBoard(board *engine.Board, history []*engine.Board) []float32 {
	encoding := make([]float32, encodingSize)

	// Channels 0-11: current position piece placement.
	encodePieces(encoding, board, 0)

	// Channel 12: side to move (entire plane 1.0 if White to move).
	if board.ActiveColor == engine.White {
		fillPlane(encoding, 12)
	}

	// Channels 13-16: castling rights.
	if board.CastlingRights&engine.CastleWhiteKing != 0 {
		fillPlane(encoding, 13)
	}
	if board.CastlingRights&engine.CastleWhiteQueen != 0 {
		fillPlane(encoding, 14)
	}
	if board.CastlingRights&engine.CastleBlackKing != 0 {
		fillPlane(encoding, 15)
	}
	if board.CastlingRights&engine.CastleBlackQueen != 0 {
		fillPlane(encoding, 16)
	}

	// Channel 17: en passant.
	if board.EnPassantSq >= 0 {
		epFile := int(board.EnPassantSq) % 8
		for rank := 0; rank < 8; rank++ {
			encoding[17*64+rank*8+epFile] = 1.0
		}
	}

	// Channels 18+: history position piece planes (12 channels each).
	// Most recent history position first.
	if len(history) > 0 {
		// Take the last numHistoryPositions entries
		start := 0
		if len(history) > numHistoryPositions {
			start = len(history) - numHistoryPositions
		}
		histSlice := history[start:]

		for i := 0; i < len(histSlice); i++ {
			// i=0 in reversed order = most recent history position
			histIdx := len(histSlice) - 1 - i
			channelOffset := currentPosChannels + i*pieceChannels
			encodePieces(encoding, histSlice[histIdx], channelOffset)
		}
	}

	return encoding
}

// encodePieces writes piece placement data for a board into 12 channels
// starting at the given channel offset.
func encodePieces(encoding []float32, board *engine.Board, channelOffset int) {
	for sq := 0; sq < 64; sq++ {
		piece := board.Squares[sq]
		if piece.IsEmpty() {
			continue
		}

		pt := piece.Type() // 1=Pawn, 2=Knight, ..., 6=King
		pieceIdx := int(pt) - 1

		var channel int
		if piece.Color() == engine.White {
			channel = channelOffset + pieceIdx
		} else {
			channel = channelOffset + 6 + pieceIdx
		}

		rank := sq / 8
		file := sq % 8
		encoding[channel*64+rank*8+file] = 1.0
	}
}

// fillPlane sets all 64 values in the given channel to 1.0.
func fillPlane(encoding []float32, channel int) {
	start := channel * 64
	for i := start; i < start+64; i++ {
		encoding[i] = 1.0
	}
}

File: rust/crates/bot/src/rl_encoder.rs
//! Board encoder for the RL agent (ported from `rl_encoder.go`).
//!
//! Produces a flat `[channel, rank, file]` float32 tensor that must stay
//! byte-identical to the Python training encoder (`training/board_encoder.py`).

use engine::{Board, Color};

const PIECE_CHANNELS: usize = 12;
const CURRENT_POS_CHANNELS: usize = 18;
const NUM_HISTORY_POSITIONS: usize = 4;
/// Total channels: 18 current-position channels + 4 history positions x 12 = 66.
pub(crate) const NUM_CHANNELS: usize =
    CURRENT_POS_CHANNELS + NUM_HISTORY_POSITIONS * PIECE_CHANNELS;
const BOARD_SIZE: usize = 8;
/// Flat encoding length: 66 * 8 * 8 = 4224.
pub(crate) const ENCODING_SIZE: usize = NUM_CHANNELS * BOARD_SIZE * BOARD_SIZE;
/// Policy vector size: 64 * 64 from-to squares = 4096.
pub(crate) const POLICY_SIZE: usize = 4096;

/// Converts a board position and its history into a 66-channel float32 tensor.
///
/// Returns a flat `Vec<f32>` of length 4224 in `[channel, rank, file]` order.
///
/// Channel layout:
/// - 0-5:   White pieces (Pawn..King) — current position
/// - 6-11:  Black pieces — current position
/// - 12:    Side to move (1.0 if White to move)
/// - 13-16: Castling rights (WK, WQ, BK, BQ)
/// - 17:    En passant file (column filled with 1.0 if available)
/// - 18-65: History piece planes (12 channels each), most recent first
pub fn encode_board(board: &Board, history: &[Board]) -> Vec<f32> {
    let mut encoding = vec![0.0f32; ENCODING_SIZE];

    // Channels 0-11: current position piece placement.
    encode_pieces(&mut encoding, board, 0);

    // Channel 12: side to move.
    if board.active_color == Color::White {
        fill_plane(&mut encoding, 12);
    }

    // Channels 13-16: castling rights.
    if board.castling_rights & engine::CASTLE_WHITE_KING != 0 {
        fill_plane(&mut encoding, 13);
    }
    if board.castling_rights & engine::CASTLE_WHITE_QUEEN != 0 {
        fill_plane(&mut encoding, 14);
    }
    if board.castling_rights & engine::CASTLE_BLACK_KING != 0 {
        fill_plane(&mut encoding, 15);
    }
    if board.castling_rights & engine::CASTLE_BLACK_QUEEN != 0 {
        fill_plane(&mut encoding, 16);
    }

    // Channel 17: en passant.
    if board.en_passant_sq >= 0 {
        let ep_file = (board.en_passant_sq as usize) % 8;
        for rank in 0..8 {
            encoding[17 * 64 + rank * 8 + ep_file] = 1.0;
        }
    }

    // Channels 18+: history position piece planes (12 channels each), most
    // recent history position first.
    if !history.is_empty() {
        let start = history.len().saturating_sub(NUM_HISTORY_POSITIONS);
        let hist_slice = &history[start..];

        for i in 0..hist_slice.len() {
            // i = 0 (reversed order) is the most recent history position.
            let hist_idx = hist_slice.len() - 1 - i;
            let channel_offset = CURRENT_POS_CHANNELS + i * PIECE_CHANNELS;
            encode_pieces(&mut encoding, &hist_slice[hist_idx], channel_offset);
        }
    }

    encoding
}

/// Writes piece placement data into 12 channels starting at `channel_offset`.
fn encode_pieces(encoding: &mut [f32], board: &Board, channel_offset: usize) {
    for sq in 0..64usize {
        let piece = board.squares[sq];
        if piece.is_empty() {
            continue;
        }

        let pt = piece.piece_type(); // 1=Pawn ... 6=King
        let piece_idx = (pt.as_u8() as usize) - 1;

        let channel = if piece.color() == Color::White {
            channel_offset + piece_idx
        } else {
            channel_offset + 6 + piece_idx
        };

        let rank = sq / 8;
        let file = sq % 8;
        encoding[channel * 64 + rank * 8 + file] = 1.0;
    }
}

/// Sets all 64 values in the given channel to 1.0.
fn fill_plane(encoding: &mut [f32], channel: usize) {
    let start = channel * 64;
    for v in encoding.iter_mut().skip(start).take(64) {
        *v = 1.0;
    }
}
